src/aspire/classification/averager2d.py

```diff
@@ -0,0 +1,1018 @@
+                assert np.sum(improved_indices) == np.size(
+                    classes
+                ), f"{np.sum(improved_indices)} =?= {np.size(classes)}"
+            else:
+                logger.debug(
+                    f"Shift ({x},{y}) complete. Improved {np.sum(improved_indices)} alignments."
+                )
+
+        return rotations, shifts, correlations
+
+
+class ReddyChatterjiAverager2D(AligningAverager2D):
+    """
+    Attempts rotational estimation using Reddy Chatterji log polar Fourier cross correlation.
+    Then attempts shift (translational) estimation using cross correlation.
+
+    When averaging, performs rotations then shifts.
+
+    Note, it may be possible to iterate this algorithm...
+
+    Adopted from Reddy Chatterji  (1996)
+    An FFT-Based Technique for Translation,
+    Rotation, and Scale-Invariant Image Registration
+    IEEE TRANSACTIONS ON IMAGE PROCESSING, VOL. 5, NO. 8, AUGUST 1996
+
+    This method intentionally does not use any of ASPIRE's basis
+    so that it may be used as a reference for more ASPIRE approaches.
+    """
+
+    def __init__(
+        self,
+        composite_basis,
+        source,
+        alignment_source=None,
+        diagnostics=False,
+        batch_size=512,
+        dtype=None,
+    ):
+        """
+        :param composite_basis:  Basis to be used during class average composition.
+        :param source: Source of original images.
+        :param alignment_source: Optional, source to be used during class average alignment.
+        Must be the same resolution as `source`.
+        :param dtype: Numpy dtype to be used during alignment.
+        """
+
+        self.__cache = dict()
+        self.diagnostics = diagnostics
+        self.do_cross_corr_translations = True
+        self.alignment_src = alignment_source or source
+
+        # TODO, for accomodating different resolutions we minimally need to adapt shifting.
+        # Outside of scope right now, but would make a nice PR later.
+        if self.alignment_src.L != source.L:
+            raise RuntimeError("Currently `alignment_src.L` must equal `source.L`")
+        if self.alignment_src.dtype != source.dtype:
+            raise RuntimeError(
+                "Currently `alignment_src.dtype` must equal `source.dtype`"
+            )
+
+        super().__init__(
+            composite_basis, source, composite_basis, batch_size=batch_size, dtype=dtype
+        )
+
+    def _phase_cross_correlation(self, img0, img1):
+        """
+        # Adapted from skimage.registration.phase_cross_correlation
+
+        :param img0: Fixed image.
+        :param img1: Translated image.
+        :returns: (cross-correlation magnitudes (2D array), shifts)
+        """
+
+        # Cache img0 transform, this saves n_classes*(n_nbor-1) transforms
+        # Note we use the `id` because ndarray are unhashable
+        src_f = self.__cache.setdefault(id(img0), fft.fft2(img0))
```

Comment:
I'm not sure what is gained here. `fft.fft2` is called regardless of whether `id(img0)` is in `__cache`.

Reply:
You're totally right, I thought this was lazy, but it was I who was lazy.  Very good catch, thank you.

Fixed.